src/base/QXmppJingleData.cpp

```diff
@@ -3224,3 +3224,314 @@ std::optional<QXmppJingleMessageInitiationElement::Type> QXmppJingleMessageIniti
+void QXmppCallInviteElement::setVideo(bool video)
+{
+    d->video = video;
+}
+
+///
+/// Returns a possible Call Invite element "jingle" sub element.
+///
+std::optional<QXmppCallInviteElement::Jingle> QXmppCallInviteElement::jingle() const
+{
+    return d->jingle;
+}
+
+///
+/// Sets a possible Call Invite "jingle" sub element.
+///
+void QXmppCallInviteElement::setJingle(std::optional<Jingle> jingle)
+{
+    d->jingle = jingle;
+}
+
+///
+/// Returns possible Call Invite "external" sub elements.
+///
+std::optional<QVector<QXmppCallInviteElement::External>> QXmppCallInviteElement::external() const
+{
+    return d->external;
+}
+
+///
+/// Sets possible Call Invite "external" sub elements.
+///
+void QXmppCallInviteElement::setExternal(std::optional<QVector<External>> external)
+{
+    d->external = external;
+}
+
+/// \cond
+void QXmppCallInviteElement::parse(const QDomElement &element)
+{
+    std::optional<Type> type { stringToCallInviteElementType(element.nodeName()) };
+
+    if (!type) {
+        return;
+    }
+
+    d->type = type.value();
+    d->id = element.attribute(QStringLiteral("id"));
+
+    switch (d->type) {
+    case Type::Invite:
+        d->audio = element.attribute(QStringLiteral("audio"), QStringLiteral("true")) == QStringLiteral("true") ? true : false;
+        d->video = element.attribute(QStringLiteral("video"), QStringLiteral("false")) == QStringLiteral("true") ? true : false;
+        // fall through
+    case Type::Accept: {
+        if (auto jingleElement = element.firstChildElement("jingle"); !jingleElement.isNull()) {
+            d->jingle = Jingle();
+            d->jingle->parse(jingleElement);
+        }
+
+        QDomElement externalItr = element.firstChildElement("external");
+
+        if (externalItr.isNull()) {
+            break;
+        }
+
+        if (!d->external.has_value()) {
+            d->external = QVector<External>();
+        }
+
+        for (; !externalItr.isNull(); externalItr = externalItr.nextSiblingElement("external")) {
+            d->external->append({ External { externalItr.attribute(QStringLiteral("uri")) } });
+        }
+
+        break;
+    }
+    case Type::Retract:
+    case Type::Reject:
+    case Type::Left:
+    default:
+        break;
+    }
+}
+
+void QXmppCallInviteElement::toXml(QXmlStreamWriter *writer) const
+{
+    // write starting tag.
+    switch (d->type) {
+    case Type::Reject:
+    case Type::Retract:
+    case Type::Left:
+        writer->writeEmptyElement(callInviteElementTypeToString(d->type));
+        break;
+    default:
+        writer->writeStartElement(callInviteElementTypeToString(d->type));
+        break;
+    }
+
+    // write namespace and ID.
+    writer->writeDefaultNamespace(ns_call_invites);
+    helperToXmlAddAttribute(writer, QStringLiteral("id"), d->id);
+
+    // no need to go on for reject, retract and left tags.
+    if (d->type == Type::Reject || d->type == Type::Retract || d->type == Type::Left) {
+        return;
+    }
```

Comment:
This code generates invalid XML, e.g. `<retract xmlns="urn:xmpp:call-invites:0"`.

Sorry that I didn't see this earlier. Please add a unit test.